docs/source/generics.rst

```diff
@@ -647,59 +693,13 @@ contravariant, use type variables defined with special keyword arguments
    my_box = Box(Square())
    look_into(my_box)  # OK, but mypy would complain here for an invariant type
 
-.. _type-variable-upper-bound:
-
-Type variables with upper bounds
-********************************
-
-A type variable can also be restricted to having values that are
-subtypes of a specific type. This type is called the upper bound of
-the type variable, and it is specified using ``T: <bound>`` when using the
-Python 3.12 syntax. In the definition of a generic function that uses
-such a type variable ``T``, the type represented by ``T`` is assumed
-to be a subtype of its upper bound, so the function can use methods
-of the upper bound on values of type ``T`` (Python 3.12 syntax):
-
-.. code-block:: python
-
-   from typing import SupportsAbs
-
-   def max_by_abs[T: SupportsAbs[float]](*xs: T) -> T:
-       # Okay, because T is a subtype of SupportsAbs[float].
-       return max(xs, key=abs)
-
-An upper bound can also be specified with the ``bound=...`` keyword
-argument to :py:class:`~typing.TypeVar`.
-Here is the example using the legacy syntax (Python 3.11 and earlier):
-
-.. code-block:: python
-
-   from typing import TypeVar, SupportsAbs
-
-   T = TypeVar('T', bound=SupportsAbs[float])
-
-   def max_by_abs(*xs: T) -> T:
-       return max(xs, key=abs)
-
-In a call to such a function, the type ``T`` must be replaced by a
-type that is a subtype of its upper bound. Continuing the example
-above:
-
-.. code-block:: python
-
-   max_by_abs(-3.5, 2)   # Okay, has type float.
-   max_by_abs(5+6j, 7)   # Okay, has type complex.
-   max_by_abs('a', 'b')  # Error: 'str' is not a subtype of SupportsAbs[float].
-
-Type parameters of generic classes may also have upper bounds, which
-restrict the valid values for the type parameter in the same way.
-
 .. _type-variable-value-restriction:
```

Comment:
Since value restrictions and bounds are closely related concepts and users are likely to confuse them, would it be better to also move up this section to just below the section on bounds?

Reply:
I don't have a strong opinion. I didn't move it since self types, which are now covered in the following section, are closely related to upper bounds (probably one of the most common use cases for upper bounds), and self types are probably much more common than constraints/value restrictions in practice.

Reply:
Sounds good, we can leave this as is.